pkg/reconciler/service/service.go

```diff
@@ -0,0 +1,482 @@
+	records := map[string]resources.RevisionRecord{}
+	// Get the list of all the revisions for the knative service.
+	revList, err := c.revisionLister.Revisions(service.Namespace).List(labels.SelectorFromSet(labels.Set{
+		serving.ConfigurationLabelKey: service.Name,
+		serving.ServiceLabelKey:       service.Name,
+	}))
+
+	if err == nil && len(revList) > 0 {
+		// Convert the list of revisions into a map of revision records, that keep the information of
+		// minScale & maxScale configured in the service, and the revision name.
+		records = CreateRevRecordsFromRevList(revList)
+	}
+	return records
+}
+
+// createRolloutOrchestrator creates the CR RolloutOrchestrator.
+func (c *Reconciler) createRolloutOrchestrator(ctx context.Context, service *servingv1.Service,
+	route *servingv1.Route) (*v1.RolloutOrchestrator, error) {
+	records := c.getRecordsFromRevs(service)
+
+	// Based on the knative service, the map of the revision records and the route, we can get the initial target
+	// revisions and the final target revisions. The initial target revisions define the start, and the final target
+	// revisions define the end for the upgrade.
+
+	// Possible use cases:
+	// 1. There is no revision records or the route, when it is the first time to create the knative service.
+	// 2. There are revision records and the route. The RolloutOrchestrator will be created on an existing old version
+	// of knative serving.
+
+	initialRevisionStatus, ultimateRevisionTarget := resources.GetInitialFinalTargetRevision(service, records, route)
+
+	// Assign the RolloutOrchestrator with the initial target revision, and final target revision.
+	// StageTargetRevisions in the spec is nil.
+	ro := resources.NewInitialFinalTargetRev(initialRevisionStatus, ultimateRevisionTarget, service)
+
+	// updateRolloutOrchestrator updates the StageRevisionTarget as the new(next) target.
+	ro = updateRolloutOrchestrator(ro, c.podAutoscalerLister.PodAutoscalers(ro.Namespace))
+	r, e := c.client.ServingV1().RolloutOrchestrators(service.Namespace).Create(
+		ctx, ro, metav1.CreateOptions{})
+
+	return r, e
+}
+
+// updateRolloutOrchestrator updates the CR RolloutOrchestrator.
+func (c *Reconciler) updateRolloutOrchestrator(ctx context.Context, service *servingv1.Service,
+	route *servingv1.Route, ro *v1.RolloutOrchestrator) (*v1.RolloutOrchestrator, error) {
+	records := c.getRecordsFromRevs(service)
+
+	// Based on the knative service, the map of the revision records and the route, we can get the final target
+	// revisions. The final target revisions define the end for the upgrade.
+	_, ultimateRevisionTarget := resources.GetInitialFinalTargetRevision(service, records, route)
+
+	// Assign the RolloutOrchestrator with the final target revision and reset StageTargetRevisions in the spec,
+	// if the final target revision is different from the existing final target revision.
+	ro = resources.UpdateFinalTargetRev(ultimateRevisionTarget, ro)
```

Comment:
Do we need to check if the previous rollout is finished before updating the final target?